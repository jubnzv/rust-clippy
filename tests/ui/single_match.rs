#![warn(clippy::single_match)]

fn dummy() {}

fn single_match() {
    let x = Some(1u8);

    match x {
        Some(y) => {
            println!("{:?}", y);
        },
        _ => (),
    };

    let x = Some(1u8);
    match x {
        // Note the missing block braces.
        // We suggest `if let Some(y) = x { .. }` because the macro
        // is expanded before we can do anything.
        Some(y) => println!("{:?}", y),
        _ => (),
    }

    let z = (1u8, 1u8);
    match z {
        (2..=3, 7..=9) => dummy(),
        _ => {},
    };

    // Not linted (pattern guards used)
    match x {
        Some(y) if y == 0 => println!("{:?}", y),
        _ => (),
    }

    // Not linted (no block with statements in the single arm)
    match z {
        (2..=3, 7..=9) => println!("{:?}", z),
        _ => println!("nope"),
    }
}

enum Foo {
    Bar,
    Baz(u8),
}
use std::borrow::Cow;
use Foo::*;

fn single_match_know_enum() {
    let x = Some(1u8);
    let y: Result<_, i8> = Ok(1i8);

    match x {
        Some(y) => dummy(),
        None => (),
    };

    match y {
        Ok(y) => dummy(),
        Err(..) => (),
    };

    let c = Cow::Borrowed("");

    match c {
        Cow::Borrowed(..) => dummy(),
        Cow::Owned(..) => (),
    };

    let z = Foo::Bar;
    // no warning
    match z {
        Bar => println!("42"),
        Baz(_) => (),
    }

    match z {
        Baz(_) => println!("42"),
        Bar => (),
    }
}

// issue #173
fn if_suggestion() {
    let x = "test";
    match x {
        "test" => println!(),
        _ => (),
    }

    #[derive(PartialEq, Eq)]
    enum Foo {
        A,
        B,
        C(u32),
    }

    let x = Foo::A;
    match x {
        Foo::A => println!(),
        _ => (),
    }

    const FOO_C: Foo = Foo::C(0);
    match x {
        FOO_C => println!(),
        _ => (),
    }

    match &&x {
        Foo::A => println!(),
        _ => (),
    }

    let x = &x;
    match &x {
        Foo::A => println!(),
        _ => (),
    }

    enum Bar {
        A,
        B,
    }
    impl PartialEq for Bar {
        fn eq(&self, rhs: &Self) -> bool {
            matches!((self, rhs), (Self::A, Self::A) | (Self::B, Self::B))
        }
    }
    impl Eq for Bar {}

    let x = Bar::A;
    match x {
        Bar::A => println!(),
        _ => (),
    }

    // issue #7038
    struct X;
    let x = Some(X);
    match x {
        None => println!(),
        _ => (),
    };
}

// See: issue #8282
fn ranges() {
    enum E {
        V,
    }
    let x = (Some(E::V), Some(42));

    // Don't lint, because the `E` enum can be extended with additional fields later. Thus, the
    // proposed replacement to `if let Some(E::V)` may hide non-exhaustive warnings that appeared
    // because of `match` construction.
    match x {
        (Some(E::V), _) => {},
        (None, _) => {},
    }
    match (Some(E::V), Some(E::V), Some(E::V)) {
        (.., Some(E::V), _) => {},
        (.., None, _) => {},
    }
    match (Some(E::V), Some(E::V), Some(E::V)) {
        (Some(E::V), ..) => {},
        (None, ..) => {},
    }
    match (Some(E::V), Some(E::V), Some(E::V)) {
        (_, Some(E::V), ..) => {},
        (_, None, ..) => {},
    }

    // Don't lint, because the second arm contains bindings. So, we cannot keep them in the `else`
    // branch.
    match (1, 2) {
        (1, _b) => (),
        (_a, ..) => (),
    }
    match (1, 2) {
        (1, _b) => {},
        (_a, _) => {},
    }

    // Lint cases.
    match x {
        (Some(_), _) => {},
        (None, _) => {},
    }
    match x {
        (Some(E::V), _) => todo!(),
        (_, _) => {},
    }
    match (Some(42), Some(E::V), Some(42)) {
        (.., Some(E::V), _) => {},
        (..) => {},
    }
}

fn annotated_bindings() {
    struct S(i32);
    let x = Some(S(1i32));

    match Some(S(1i32)) {
        Some(a) => {},
        _ => (),
    }

    match x {
        Some(ref a) => {},
        _ => (),
    }

    match x {
        Some(mut a) => {},
        _ => (),
    }

    match Some(S(1i32)) {
        Some(ref mut a) => {},
        _ => (),
    }
}

#[allow(clippy::redundant_pattern)]
fn binding_subpatterns() {
    // Lint.
    match 1 {
        _a @ 42 => {},
        _ => {},
    }
    match 1 {
        _a @ 42 => {},
        _a @ _ => {},
    }

    // Don't lint.
    match 1 {
        _a @ 42 => {},
        _b => {},
    }
}

fn tuple_structs() {
    struct S(i32, i32);
    let s = S(1, 2);

    // dont' lint
    match s {
        S(42, _a) => {},
        S(_, _) => {},
    }

    // don't lint
    match s {
        S(42, _a) => {},
        S(..) => {},
    }

    enum E {
        S1(i32, i32),
        S2(i32, i32),
    }
    let s = E::S1(1, 2);

    // don't lint
    match s {
        E::S1(..) => {},
        E::S2(..) => {},
    }
}

fn structs() {
    struct S {
        x: i32,
        y: i32,
    }
    let s = S { x: 1, y: 2 };

    // lint
    match s {
        S { x: 42, .. } => {},
        S { .. } => {},
    }

    // lint
    match s {
        S { x: _x, y: 42 } => {},
        S { .. } => {},
    }

    // don't lint
    match s {
        S { x: i32::MIN, y: _y } => {},
        S {
            x: i32::MIN..=i32::MAX, ..
        } => {},
    }

    // Don't lint, because we have an annotated binding in the second arm. So we cannot replace it
    // with `else` branch, keeping this binding.
    match s {
        S { x: _x, y: _y } => dummy(),
        S { x: _x, .. } => dummy(),
    }
}

macro_rules! single_match {
    ($num:literal) => {
        match $num {
            15 => println!("15"),
            _ => (),
        }
    };
}

fn main() {
    single_match!(5);
}
