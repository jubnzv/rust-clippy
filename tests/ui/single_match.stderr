error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:8:5
   |
LL | /     match x {
LL | |         Some(y) => {
LL | |             println!("{:?}", y);
LL | |         },
LL | |         _ => (),
LL | |     };
   | |_____^
   |
   = note: `-D clippy::single-match` implied by `-D warnings`
help: try this
   |
LL ~     if let Some(y) = x {
LL +         println!("{:?}", y);
LL ~     };
   |

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:16:5
   |
LL | /     match x {
LL | |         // Note the missing block braces.
LL | |         // We suggest `if let Some(y) = x { .. }` because the macro
LL | |         // is expanded before we can do anything.
LL | |         Some(y) => println!("{:?}", y),
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if let Some(y) = x { println!("{:?}", y) }`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:25:5
   |
LL | /     match z {
LL | |         (2..=3, 7..=9) => dummy(),
LL | |         _ => {},
LL | |     };
   | |_____^ help: try this: `if let (2..=3, 7..=9) = z { dummy() }`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:59:5
   |
LL | /     match y {
LL | |         Ok(y) => dummy(),
LL | |         Err(..) => (),
LL | |     };
   | |_____^ help: try this: `if let Ok(y) = y { dummy() }`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:66:5
   |
LL | /     match c {
LL | |         Cow::Borrowed(..) => dummy(),
LL | |         Cow::Owned(..) => (),
LL | |     };
   | |_____^ help: try this: `if let Cow::Borrowed(..) = c { dummy() }`

error: you seem to be trying to use `match` for an equality check. Consider using `if`
  --> $DIR/single_match.rs:87:5
   |
LL | /     match x {
LL | |         "test" => println!(),
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if x == "test" { println!() }`

error: you seem to be trying to use `match` for an equality check. Consider using `if`
  --> $DIR/single_match.rs:100:5
   |
LL | /     match x {
LL | |         Foo::A => println!(),
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if x == Foo::A { println!() }`

error: you seem to be trying to use `match` for an equality check. Consider using `if`
  --> $DIR/single_match.rs:106:5
   |
LL | /     match x {
LL | |         FOO_C => println!(),
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if x == FOO_C { println!() }`

error: you seem to be trying to use `match` for an equality check. Consider using `if`
  --> $DIR/single_match.rs:111:5
   |
LL | /     match &&x {
LL | |         Foo::A => println!(),
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if x == Foo::A { println!() }`

error: you seem to be trying to use `match` for an equality check. Consider using `if`
  --> $DIR/single_match.rs:117:5
   |
LL | /     match &x {
LL | |         Foo::A => println!(),
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if x == &Foo::A { println!() }`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:134:5
   |
LL | /     match x {
LL | |         Bar::A => println!(),
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if let Bar::A = x { println!() }`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:142:5
   |
LL | /     match x {
LL | |         None => println!(),
LL | |         _ => (),
LL | |     };
   | |_____^ help: try this: `if let None = x { println!() }`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:164:5
   |
LL | /     match x {
LL | |         (Some(_), _) => {},
LL | |         (None, _) => {},
LL | |     }
   | |_____^ help: try this: `if let (Some(_), _) = x {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:170:5
   |
LL | /     match x {
LL | |         (Some(E::V), _) => todo!(),
LL | |         (_, _) => {},
LL | |     }
   | |_____^ help: try this: `if let (Some(E::V), _) = x { todo!() }`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:176:5
   |
LL | /     match (Some(42), Some(E::V), Some(42)) {
LL | |         (.., Some(E::V), _) => {},
LL | |         (..) => {},
LL | |     }
   | |_____^ help: try this: `if let (.., Some(E::V), _) = (Some(42), Some(E::V), Some(42)) {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:204:5
   |
LL | /     match Some(S(1i32)) {
LL | |         Some(a) => {},
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if let Some(a) = Some(S(1i32)) {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:209:5
   |
LL | /     match x {
LL | |         Some(ref a) => {},
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if let Some(ref a) = x {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:214:5
   |
LL | /     match x {
LL | |         Some(mut a) => {},
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if let Some(mut a) = x {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:219:5
   |
LL | /     match Some(S(1i32)) {
LL | |         Some(ref mut a) => {},
LL | |         _ => (),
LL | |     }
   | |_____^ help: try this: `if let Some(ref mut a) = Some(S(1i32)) {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:228:5
   |
LL | /     match x {
LL | |         S(aa @ 42) => {},
LL | |         S(_) => {},
LL | |     }
   | |_____^ help: try this: `if let S(aa @ 42) = x {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:239:5
   |
LL | /     match s {
LL | |         S(42, _a) => {},
LL | |         S(_, _) => {},
LL | |     }
   | |_____^ help: try this: `if let S(42, _a) = s {}`

error: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> $DIR/single_match.rs:245:5
   |
LL | /     match s {
LL | |         S(42, _a) => {},
LL | |         S(..) => {},
LL | |     }
   | |_____^ help: try this: `if let S(42, _a) = s {}`

error: aborting due to 22 previous errors

